From 9d11bc6ef3fd5889c8fa7a21e0f6a8df8cf011c1 Mon Sep 17 00:00:00 2001
From: Joan Torres Lopez <joantolo@redhat.com>
Date: Tue, 3 Mar 2026 17:57:16 +0100
Subject: [PATCH] daemon-system: Respect existing remote_id from third-party
 remote displays

When a remote display already has a valid remote_id set (not "/" or null),
reuse that id instead of generating a new one and overwriting it. This
allows third-party remote desktop solutions to work correctly by preserving
their original remote_id. Only generate and set a new remote_id when the
display doesn't have one.

Part-of: <https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/merge_requests/386>
---
 src/grd-daemon-system.c | 46 +++++++++++++++++++++++++++++++++++------
 1 file changed, 40 insertions(+), 6 deletions(-)

diff --git a/src/grd-daemon-system.c b/src/grd-daemon-system.c
index 5c7f7ec..d5c382a 100644
--- a/src/grd-daemon-system.c
+++ b/src/grd-daemon-system.c
@@ -1092,17 +1092,32 @@ on_gdm_remote_display_session_id_changed (GrdDBusGdmRemoteDisplay *remote_displa
   register_handover_iface (remote_client, session_id);
 }
 
-static GrdRemoteClient *
-remote_client_new_from_display (GrdDaemonSystem         *daemon_system,
-                                GrdDBusGdmRemoteDisplay *remote_display)
+static char *
+get_remote_id_from_remote_display (GrdDaemonSystem         *daemon_system,
+                                   GrdDBusGdmRemoteDisplay *remote_display)
 {
-  g_autoptr (GrdRemoteClient) remote_client = NULL;
+  const char *display_remote_id;
+  g_autofree char *new_remote_id = NULL;
   g_autoptr (GError) error = NULL;
 
-  remote_client = remote_client_new (daemon_system, NULL);
+  display_remote_id = grd_dbus_gdm_remote_display_get_remote_id (remote_display);
+  if (display_remote_id && !g_str_equal (display_remote_id, "/"))
+    {
+      if (g_hash_table_contains (daemon_system->remote_clients, display_remote_id))
+        {
+          g_warning ("[DaemonSystem] A remote client with remote_id %s already "
+                     "exists", display_remote_id);
+          return NULL;
+        }
+
+      return g_strdup (display_remote_id);
+    }
 
+  /* If the remote display doesn't have a valid remote_id,
+   * generate a new one and set it */
+  new_remote_id = get_next_available_id (daemon_system);
   if (!grd_dbus_gdm_remote_display_call_set_remote_id_sync (remote_display,
-                                                            remote_client->id,
+                                                            new_remote_id,
                                                             NULL,
                                                             &error))
     {
@@ -1111,6 +1126,25 @@ remote_client_new_from_display (GrdDaemonSystem         *daemon_system,
       return NULL;
     }
 
+  return g_steal_pointer (&new_remote_id);
+}
+
+static GrdRemoteClient *
+remote_client_new_from_display (GrdDaemonSystem         *daemon_system,
+                                GrdDBusGdmRemoteDisplay *remote_display)
+{
+  g_autoptr (GrdRemoteClient) remote_client = NULL;
+  char *remote_id;
+
+  remote_client = g_new0 (GrdRemoteClient, 1);
+  remote_client->daemon_system = daemon_system;
+
+  remote_id = get_remote_id_from_remote_display (daemon_system, remote_display);
+  if (!remote_id)
+    return NULL;
+
+  remote_client->id = remote_id;
+
   return g_steal_pointer (&remote_client);
 }
 
-- 
2.51.0

